laurent22/joplin
# Pr/mobile plugin support/just panels and dialogs

# Summary

This pull request does two things:
1. Adds support for plugin dialogs, panels, and more commands
2. Adds support for the Outline plugin -- Adds polyfills, mocks, and fixes a `supportsCommand` bug.

# Screenshots

![screenshot: "Show plugin panels" menu item highlighted](https://github.com/laurent22/joplin/assets/46334387/a1caa035-f258-4ef2-8c73-4bd7c9b5e0c7)

# Notes

I'm not sure how much I like the current way to show and hide plugin panels. The note actions menu is getting very full.

Here are a few other possible GUIs:

## 1. Show the plugin panels in the "properties" sidebar

![screenshot: Sidebar on right side of screen expanded, includes created and updated time, and view on map. Just below view on map is a large blue rectangle.](https://github.com/laurent22/joplin/assets/46334387/89e6fe9a-b2d4-4e43-9e30-25677ee257d8)

## 2. Show the toggle button in the main toolbar

This has the benefit of allowing plugins to be shown even when a note is not open. I also have a draft implementation of this on a separate branch (though it needs accessibility changes).

![screenshot: main toolbar toggle button shown, clicking on it (indicated by an arrow), shows a dialog that lists open plugin panels](https://github.com/laurent22/joplin/assets/46334387/70a7ccad-4654-4471-bafd-f2c0a8516a85)


# Testing

To test this pull request,
1. Install the "Outline" and "automatic backlinks" plugins (and also possibly "Link Graph UI")
2. Enable the "automatic backlinks" panel in its settings
3. Open a note
4. Show the plugin panels
5. Switch between them
6. Verify that backlinks and headings are shown in the outline and backlinks panels, respectively.

This has been tested manually on an Android 12 emulator.
